groups/bsl/bslstl/bslstl_optional.h

```diff
@@ -0,0 +1,1611 @@
+// variable below not to be extern.  The variable itself requires a constructor
+// definition before its own definition.
+inline
+BSLS_KEYWORD_CONSTEXPR nullopt_t::nullopt_t(int) BSLS_KEYWORD_NOEXCEPT
+{
+}
+
+static const BSLS_KEYWORD_CONSTEXPR nullopt_t nullopt = nullopt_t(0);
+// Value of type 'nullopt_t' used as an argument to functions that take a
+// 'nullopt_t' argument.
+
+#endif  // __cpp_lib_optional
+
+template <class TYPE,
+          bool USES_BSLMA_ALLOC =
+              BloombergLP::bslma::UsesBslmaAllocator<TYPE>::value>
+class optional;
+
+}  // close namespace bsl
+
+namespace BloombergLP {
+namespace bslstl {
+
+                        // ============================
+                        // class Optional_OptNoSuchType
+                        // ============================
+
+struct Optional_OptNoSuchType {
+    // This trivial tag type is used to distinguish between arguments passed by
+    // a user, and an 'enable_if' argument.  It is not default constructible so
+    // the following construction never invokes a constrained single parameter
+    // constructor:
+    //..
+    //   optional<SomeType> o(int, {});
+    //..
+
+    // CREATORS
+    explicit BSLS_KEYWORD_CONSTEXPR Optional_OptNoSuchType(
+                                                    int) BSLS_KEYWORD_NOEXCEPT;
+        // Create an 'Optional_OptNoSuchType' object.  Note that the argument is
+        // not used.
+};
+
+// CREATORS
+
+// bde_verify requires an out-of-class definition.  It also requires the tag
+// variable below not to be extern.  The variable itself requires a constructor
+// definition before its own definition.
+inline
+BSLS_KEYWORD_CONSTEXPR Optional_OptNoSuchType::Optional_OptNoSuchType(
+                                                     int) BSLS_KEYWORD_NOEXCEPT
+{
+}
+
+static const BSLS_KEYWORD_CONSTEXPR Optional_OptNoSuchType optNoSuchType =
+    Optional_OptNoSuchType(0);
+// Value of type 'Optional_OptNoSuchType' used as the default argument in
+// functions that are constrained using a function argument.
+
+#ifdef BSLS_LIBRARYFEATURES_HAS_CPP11_BASELINE_LIBRARY
+#define BSLSTL_OPTIONAL_IS_TRIVIALLY_DESTRUCTIBLE                             \
+    std::is_trivially_destructible
+
+#else
+#define BSLSTL_OPTIONAL_IS_TRIVIALLY_DESTRUCTIBLE bsl::is_trivially_copyable
+// C++03 does not provide a trivially destructible state.  Instead we use
+// 'bsl::is_trivially_copyable' which implies the type is also trivially
+// destructible.
+#endif  // BSLS_LIBRARYFEATURES_HAS_CPP11_BASELINE_LIBRARY
+
+                           // ======================
+                           // class Optional_DataImp
+                           // ======================
+
+template <class TYPE>
+struct Optional_DataImp {
+    // This component-private 'struct' manages a 'value_type' object in
+    // 'optional'.  It allows 'optional<TYPE>' to be trivially destructible if
+    // 'TYPE' is trivially destructible.  This class also provides an
```

Comment:
This one doesn't handle triviality, right?
```suggestion
    // 'optional'.  This class provides an
```

Reply:
correct, this doesn't belong here